src/Reflector/Generic.php

```diff
@@ -0,0 +1,196 @@
+     * @return Reflection|null
+     */
+    private function reflectNode(Node $node, Node\Stmt\Namespace_ $namespace = null, $filename = null)
+    {
+        if ($node instanceof Node\Stmt\Class_) {
+            return ReflectionClass::createFromNode(
+                $node,
+                $namespace,
+                $filename
+            );
+        }
+
+        return null;
+    }
+
+    /**
+     * Process and reflect all the matching symbols found inside a namespace node
+     *
+     * @param Node\Stmt\Namespace_ $namespace
+     * @param Symbol $symbol
+     * @param string|null $filename
+     * @return Reflection[]
+     */
+    private function reflectFromNamespace(
+        Node\Stmt\Namespace_ $namespace,
+        Symbol $symbol,
+        $filename
+    ) {
+        $reflections = [];
+        foreach ($namespace->stmts as $node) {
+            $reflection = $this->reflectNode($node, $namespace, $filename);
+
+            if (null !== $reflection && $symbol->isMatchingReflector($reflection)) {
+                $reflections[] = $reflection;
+            }
+
+        }
+        return $reflections;
+    }
+
+    /**
+     * Reflect symbols from an AST. If a namespace is found, also load all the
+     * matching symbols found in the namespace
+     *
+     * @param Node[] $ast
+     * @param string|null $filename
+     * @param Symbol $symbol
+     * @return Reflection[]
+     */
+    private function reflectFromTree(array $ast, $filename, Symbol $symbol)
+    {
+        $reflections = [];
+        foreach ($ast as $node) {
+            switch (get_class($node)) {
+                case Node\Stmt\Namespace_::class:
+                    $reflections = array_merge(
+                        $reflections,
+                        $this->reflectFromNamespace($node, $symbol, $filename)
+                    );
+                    break;
+                case Node\Stmt\Class_::class:
+                    if ($symbol->getType() == Symbol::SYMBOL_CLASS) {
+                        $reflections[] = $this->reflectNode($node, null, $filename);
+                    }
+                    break;
+            }
+        }
+        return $reflections;
+    }
+
+    /**
+     * Get an array of reflections found in a LocatedSource
+     *
+     * @param LocatedSource $locatedSource
+     * @param Symbol $symbol
+     * @return Reflection[]
+     */
+    private function getReflections(LocatedSource $locatedSource, Symbol $symbol)
+    {
+        $parser = new Parser(new Lexer);
+        $ast = $parser->parse($locatedSource->getSource());
+
+        return $this->reflectFromTree(
+            $ast,
+            $locatedSource->getFileName(),
+            $symbol
+        );
+    }
+
+    /**
+     * Get all symbols of a matching symbol type from a file
+     *
+     * @param string $symbolType
+     * @return Reflection[]
+     */
+    public function getAllSymbols($symbolType)
+    {
+        $symbol = new Symbol('*', $symbolType);
```

Comment:
it actually doesn't matter what it is - or at least in the case of SingleFileSourceLocator it didn't - I need to check this